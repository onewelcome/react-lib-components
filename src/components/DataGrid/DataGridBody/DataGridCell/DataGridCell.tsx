/*
 * Copyright 2022 OneWelcome B.V.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

import React, {
  ForwardRefRenderFunction,
  ComponentPropsWithRef,
  ReactElement,
  Fragment
} from "react";
import { Typography } from "../../../Typography/Typography";
import classes from "./DataGridCell.module.scss";

export interface Props extends ComponentPropsWithRef<"td"> {
  children?: ReactElement | string | number;
  prefixElement?: ReactElement;
  isLoading?: boolean;
  spacing?: React.CSSProperties;
  searchValue?: string;
  cellIndex?: number;
  columnLength?: number;
  disableContextMenuColumn?: boolean;
}

const DataGridCellComponent: ForwardRefRenderFunction<HTMLTableCellElement, Props> = (
  {
    children,
    className,
    isLoading,
    spacing,
    prefixElement,
    searchValue,
    cellIndex,
    columnLength,
    disableContextMenuColumn,
    ...rest
  }: Props,
  ref
) => {
  let cellStyle: React.CSSProperties = {};

  if (cellIndex === 0) {
    cellStyle.paddingLeft = spacing?.paddingLeft;
  }
  if (
    (cellIndex === columnLength && !disableContextMenuColumn) ||
    (columnLength && cellIndex === columnLength - 1 && disableContextMenuColumn)
  ) {
    cellStyle.paddingRight = spacing?.paddingRight;
  }

  //NOTE: we might want to migrate to Highlight API once it's supported by Firefox
  const renderContent = () => {
    if (typeof children === "string" && searchValue) {
      if (!children.toLowerCase().includes(searchValue.toLowerCase())) {
        return children;
      }

      const matchingSequence = new RegExp(searchValue, "i").exec(children);

      const parts = children.split(matchingSequence?.[0] ?? "");
      return parts.map((part, i) => {
        if (i === parts.length - 1) return <Fragment key={`${part}-${i}`}>{part}</Fragment>;

        return (
          <Fragment key={`${part}-${i}`}>
            {part}
            <mark data-testid={`${matchingSequence}-mark`}>{matchingSequence}</mark>
          </Fragment>
        );
      });
    }

    return children;
  };

  return (
    <td
      {...rest}
      ref={ref}
      style={{ ...rest.style, ...cellStyle }}
      className={`${classes["cell"]} ${className ?? ""}`}
    >
      {prefixElement ? (
        <div className={classes["content-wrapper"]}>
          {prefixElement}
          {isLoading && (
            <div className={classes["loading"]} aria-busy="true" aria-live="polite"></div>
          )}
          {!isLoading && (
            <Typography
              className={classes["text"]}
              variant="body"
              tag="span"
              spacing={{ marginBottom: 0 }}
            >
              {renderContent()}
            </Typography>
          )}
        </div>
      ) : (
        <>
          {isLoading && (
            <div className={classes["loading"]} aria-busy="true" aria-live="polite"></div>
          )}
          {!isLoading && (
            <Typography className={classes["text"]} variant="body" tag="span">
              {renderContent()}
            </Typography>
          )}
        </>
      )}
    </td>
  );
};

export const DataGridCell = React.forwardRef(DataGridCellComponent);
