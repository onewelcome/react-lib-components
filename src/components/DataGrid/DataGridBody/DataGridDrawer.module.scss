.drawer {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
  padding: 1rem 4.75rem;
}
