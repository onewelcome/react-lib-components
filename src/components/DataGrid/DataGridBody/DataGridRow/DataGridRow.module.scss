/*!
 * Copyright 2022 OneWelcome B.V.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

.row {
  position: relative;
  background-color: var(--data-grid-row-background-color);

  &:not(.loading):hover {
    background-color: var(--data-grid-row-hover-background-color);
  }

  @mixin borderAfter {
    content: "";
    position: absolute;
    width: 100%;
    left: 0;
    bottom: 0;
    border-bottom: 1px solid var(--light-grey-border);
  }

  &.border {
    &:not(:last-child)::after {
      @include borderAfter();
    }
  }

  &.border-drawer {
    &:not(:last-child):not(:nth-last-child(2))::after {
      @include borderAfter();
    }
  }
}

.expand-button-cell {
  padding: 0;
}
