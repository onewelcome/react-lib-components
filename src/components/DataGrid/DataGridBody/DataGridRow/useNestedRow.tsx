/*
 * Copyright 2022 OneWelcome B.V.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

import React, { ReactElement } from "react";
import classes from "./DataGridRow.module.scss";
import { HeaderCell } from "../../datagrid.interfaces";

export interface UseNestedRowProps<T> {
  indentationLevels?: { level: number; isLastChild: boolean }[];
  item?: T;
  nestedItemsKey?: keyof T;
  rowTemplate?: ({ item, index }: { item: T; index: number }) => ReactElement;
  isRowExpanded?: boolean;
  indentationLevel: number;
  enableNestedRow?: boolean;
  rowProps: {
    searchValue?: string;
    headers?: HeaderCell[];
    spacing?: React.CSSProperties;
    disableContextMenuColumn?: boolean;
  };
}

export const useNestedRow = <T,>({
  indentationLevels,
  item,
  nestedItemsKey,
  rowTemplate,
  isRowExpanded,
  indentationLevel,
  enableNestedRow,
  rowProps: { searchValue, headers, spacing, disableContextMenuColumn }
}: UseNestedRowProps<T>) => {
  const hasNestedChildren = item && nestedItemsKey && item[nestedItemsKey];

  const getNestedChildSpacing = (spacing: React.CSSProperties | undefined, index: number) => {
    const expandButtonWidth = 46;
    const nestedChildOffset = !hasNestedChildren ? expandButtonWidth : 0;

    const connectorWidth = 68;
    const nestedChildIndentation = `${nestedChildOffset + indentationLevel * connectorWidth}`;

    const gap = 4;
    const notIndentedWithNoChildrenOffset =
      indentationLevel === 0 && !hasNestedChildren ? expandButtonWidth + gap : gap;
    const childIndentation = `${indentationLevel ? nestedChildIndentation : notIndentedWithNoChildrenOffset}px`;

    if (spacing) {
      return {
        ...spacing,
        paddingLeft: index === 0 ? childIndentation : spacing.paddingLeft
      };
    }
    return {
      paddingLeft: index === 0 ? childIndentation : ""
    };
  };

  const renderNestedRowConnectors = () => {
    if (indentationLevels) {
      return indentationLevels.map(({ level, isLastChild }) =>
        renderNestedRowConnector(level, isLastChild, indentationLevels.length)
      );
    }

    return null;
  };

  const renderNestedRowConnector = (level: number, isLastChild: boolean, levelsLength: number) => {
    const offsetLeftClass = classes[`offset-left-${level - 1}`];

    if (level === levelsLength) {
      const variant = isLastChild ? "line" : "t-shape";

      return (
        <div
          key={level}
          data-testid="dataGridRowConnector"
          className={`${classes["connector"]} ${classes[variant]} ${offsetLeftClass}`}
        />
      );
    } else if (!isLastChild) {
      return (
        <div
          key={level}
          data-testid="dataGridRowConnector"
          className={`${classes["connector"]} ${classes["vertical"]}  ${offsetLeftClass}`}
        />
      );
    }
    return null;
  };

  const renderNestedRow = () => {
    if (rowTemplate && item && nestedItemsKey && item[nestedItemsKey] && isRowExpanded) {
      const nestedItemsArray: T[] = item[nestedItemsKey] as T[];
      const getIndentationLevel = (index: number) => ({
        level: indentationLevel + 1,
        isLastChild: index + 1 === nestedItemsArray.length
      });

      return nestedItemsArray.map((item, index) => {
        return React.cloneElement(rowTemplate({ item, index }), {
          searchValue,
          headers,
          spacing,
          disableContextMenuColumn,
          item,
          nestedRowProps: {
            rowTemplate,
            enableNestedRow,
            nestedItemsKey,
            indentationLevel: indentationLevel + 1,
            indentationLevels:
              indentationLevels && nestedItemsArray
                ? [...indentationLevels, getIndentationLevel(index)]
                : [getIndentationLevel(index)]
          }
        });
      });
    }
  };

  return {
    renderNestedRow,
    renderNestedRowConnectors,
    getNestedChildSpacing,
    hasNestedChildren
  };
};
