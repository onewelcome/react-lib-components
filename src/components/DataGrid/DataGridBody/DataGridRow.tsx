/*
 * Copyright 2022 OneWelcome B.V.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

import React, {
  ComponentPropsWithRef,
  useState,
  Fragment,
  Ref,
  ReactElement,
  ForwardRefRenderFunction
} from "react";
import { HeaderCell } from "../datagrid.interfaces";
import classes from "./DataGridRow.module.scss";
import { IconButton } from "../../Button/IconButton";
import { Icon, Icons } from "../../Icon/Icon";
import { DataGridCell } from "./DataGridCell";
import { DataGridDrawer } from "./DataGridDrawer";
import { generateID } from "../../../util/helper";

export interface Props extends ComponentPropsWithRef<"tr"> {
  expandButtonTitle?: string;
  expandButtonId?: string;
  drawerId?: string;
  headers?: HeaderCell[];
  isLoading?: boolean;
  spacing?: React.CSSProperties;
  disableContextMenuColumn?: boolean;
  enableExpandableRow?: boolean;
  expandableRowContent?: React.ReactNode;
  rowExpanded?: boolean;
  onExpandRowButtonClick?: boolean;
}

const DataGridRowComponent: ForwardRefRenderFunction<HTMLTableRowElement, Props> = (
  {
    expandButtonTitle = "Expand row",
    expandButtonId = `ID-${generateID()}`,
    drawerId = `ID-${generateID()}`,
    children,
    className,
    headers,
    isLoading,
    spacing,
    expandableRowContent,
    disableContextMenuColumn,
    enableExpandableRow,
    ...rest
  },
  ref
) => {
  const [isRowExpanded, setIsRowExpanded] = useState(false);
  const visibleCells = React.Children.map(children as React.ReactElement[], (child, index) => {
    if (child) {
      const cellWithSpacing = React.cloneElement(child, {
        spacing: spacing,
        cellIndex: index,
        columnLength: headers?.length,
        disableContextMenuColumn
      });

      const visible = headers && headers.length > index ? !headers[index].hidden : true;
      return visible && cellWithSpacing;
    }
    return null;
  });

  const classNames = [classes["row"]];
  className && classNames.push(className);
  !isRowExpanded && classNames.push(classes["border"]);
  isLoading && classNames.push(classes["loading"]);

  return (
    <Fragment>
      <tr {...rest} ref={ref} className={classNames.join(" ")}>
        {enableExpandableRow && (
          <DataGridCell
            className={classes["expand-button-cell"]}
            onClick={() => setIsRowExpanded(!isRowExpanded)}
            style={{ width: "1px" }}
          >
            <IconButton
              id={expandButtonId}
              title={expandButtonTitle}
              aria-expanded={isRowExpanded}
              aria-controls={drawerId}
            >
              <Icon size="0.75rem" icon={isRowExpanded ? Icons.ChevronUp : Icons.ChevronDown} />
            </IconButton>
          </DataGridCell>
        )}
        {visibleCells}
      </tr>
      {enableExpandableRow && (
        <tr className={`${classes["row"]} ${isRowExpanded ? classes["border"] : ""}`}>
          <td colSpan={visibleCells.length + 1}>
            <DataGridDrawer
              id={drawerId}
              role="region"
              aria-labelledby={expandButtonId}
              isExpanded={isRowExpanded}
            >
              {expandableRowContent}
            </DataGridDrawer>
          </td>
        </tr>
      )}
    </Fragment>
  );
};

export const DataGridRow = React.forwardRef(DataGridRowComponent);
