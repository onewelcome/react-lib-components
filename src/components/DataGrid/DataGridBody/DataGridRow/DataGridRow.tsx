/*
 * Copyright 2022 OneWelcome B.V.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

import React, { ComponentPropsWithRef, useState, Fragment, Ref, ReactElement } from "react";
import { HeaderCell } from "../../datagrid.interfaces";
import classes from "./DataGridRow.module.scss";
import { IconButton } from "../../../Button/IconButton";
import { Icon, Icons } from "../../../Icon/Icon";
import { DataGridCell } from "../DataGridCell/DataGridCell";
import { DataGridDrawer } from "../DataGridDrawer/DataGridDrawer";
import { generateID } from "../../../../util/helper";

export interface Props<T> extends ComponentPropsWithRef<"tr"> {
  item?: T;
  indentationLevel?: number;
  indentationLevels?: { level: number; isLastChild: boolean }[];
  rowTemplate?: ({ item, index }: { item: T; index: number }) => ReactElement;
  headers?: HeaderCell[];
  isLoading?: boolean;
  spacing?: React.CSSProperties;
  searchValue?: string;
  disableContextMenuColumn?: boolean;
  enableNestedRow?: boolean;
  isLastRootAncestor?: boolean;
  isLastChild?: boolean;
  nestedItemsKey?: keyof T;
  expandableRowProps?: {
    enableExpandableRow: boolean;
    expandableRowContent: React.ReactNode;
    expandButtonTitle?: string;
    expandButtonId?: string;
    drawerId?: string;
  };
}

const DataGridRowComponent = <T,>(
  {
    item,
    children,
    className,
    rowTemplate,
    headers,
    searchValue,
    isLoading,
    spacing,
    expandableRowProps,
    disableContextMenuColumn,
    enableNestedRow,
    indentationLevel = 0,
    indentationLevels,
    isLastChild,
    nestedItemsKey,
    ...rest
  }: Props<T>,
  ref: Ref<HTMLTableRowElement>
) => {
  const {
    enableExpandableRow = false,
    expandButtonId = `ID-${generateID()}`,
    expandButtonTitle = "Expand row",
    drawerId = `ID-${generateID()}`,
    expandableRowContent
  } = expandableRowProps || {};
  const [isRowExpanded, setIsRowExpanded] = useState(false);

  const classNames = [classes["row"]];
  const rowBorderClass = enableNestedRow
    ? classes[`border-${indentationLevel}`]
    : classes[`border`];

  className && classNames.push(className);
  enableExpandableRow
    ? !isRowExpanded && classNames.push(classes["border-drawer"])
    : classNames.push(rowBorderClass);
  isLoading && classNames.push(classes["loading"]);

  const renderNestedRowConnectors = () => {
    if (indentationLevels) {
      return indentationLevels.map(({ level, isLastChild }) =>
        renderNestedRowConnector(level, isLastChild, indentationLevels.length)
      );
    }

    return null;
  };

  const renderNestedRowConnector = (level: number, isLastChild: boolean, levelsLength: number) => {
    const offsetLeftClass = classes[`offset-left-${level - 1}`];

    if (level === levelsLength) {
      const variant = isLastChild ? "line" : "t-shape";

      return (
        <div
          data-testid="dataGridRowConnector"
          className={`${classes["connector"]} ${classes[variant]} ${offsetLeftClass}`}
        />
      );
    } else if (!isLastChild) {
      return (
        <div
          data-testid="dataGridRowConnector"
          className={`${classes["connector"]} ${classes["vertical"]}  ${offsetLeftClass}`}
        />
      );
    }
    return null;
  };

  const renderRecurrentRow = () => {
    if (rowTemplate && item && nestedItemsKey && item[nestedItemsKey] && isRowExpanded) {
      const nestedItemsArray: T[] = item[nestedItemsKey] as T[];
      const getIndentationLevel = (index: number) => ({
        level: indentationLevel + 1,
        isLastChild: index + 1 === nestedItemsArray.length
      });

      return nestedItemsArray.map((item, index) => {
        return React.cloneElement(rowTemplate({ item, index }), {
          searchValue: searchValue,
          headers,
          spacing,
          disableContextMenuColumn,
          enableNestedRow,
          nestedItemsKey,
          indentationLevel: indentationLevel + 1,
          indentationLevels:
            indentationLevels && nestedItemsArray
              ? [...indentationLevels, getIndentationLevel(index)]
              : [getIndentationLevel(index)],
          item,
          rowTemplate,
          isLastChild: nestedItemsArray ? index + 1 === nestedItemsArray.length : false
        });
      });
    }
  };

  const visibleCells = React.Children.map(children as React.ReactElement[], (child, index) => {
    const hasNestedChildren = item && nestedItemsKey && item[nestedItemsKey];
    const nestedChildOffset = !hasNestedChildren ? 46 : 0;
    const nestedChildIndentation = `${nestedChildOffset + indentationLevel * 68}`;
    const notIndentedWithNoChildrenOffset = indentationLevel === 0 && !hasNestedChildren ? 50 : 4;
    const childIndentation = `${indentationLevel ? nestedChildIndentation : notIndentedWithNoChildrenOffset}px`;

    const getNestedChildSpacing = (spacing: React.CSSProperties | undefined) => {
      if (spacing) {
        return {
          ...spacing,
          paddingLeft: index === 0 ? childIndentation : spacing.paddingLeft
        };
      }
      return {
        paddingLeft: index === 0 ? childIndentation : ""
      };
    };

    const childSpacing = enableNestedRow ? getNestedChildSpacing(spacing) : spacing;

    const prefixButton = hasNestedChildren ? (
      <IconButton
        id={expandButtonId}
        title={expandButtonTitle}
        aria-expanded={isRowExpanded}
        onClick={() => setIsRowExpanded(!isRowExpanded)}
      >
        <Icon size="0.75rem" icon={isRowExpanded ? Icons.ChevronUp : Icons.ChevronDown} />
      </IconButton>
    ) : null;

    const prefixElement =
      enableNestedRow && index === 0 ? (
        <>
          {prefixButton}
          {renderNestedRowConnectors()}
        </>
      ) : null;

    if (child) {
      const cellWithSpacing = React.cloneElement(child, {
        searchValue,
        spacing: childSpacing,
        cellIndex: index,
        columnLength: headers?.length,
        disableContextMenuColumn,
        prefixElement
      });

      const visible = headers && headers.length > index ? !headers[index].hidden : true;
      return visible && cellWithSpacing;
    }
    return null;
  });

  return (
    <Fragment>
      <tr {...rest} ref={ref} className={classNames.join(" ")}>
        {enableExpandableRow && (
          <DataGridCell
            className={classes["expand-button-cell"]}
            onClick={() => setIsRowExpanded(!isRowExpanded)}
            style={{
              width: "1px"
            }}
          >
            <IconButton
              id={expandButtonId}
              title={expandButtonTitle}
              aria-expanded={isRowExpanded}
              aria-controls={drawerId}
            >
              <Icon size="0.75rem" icon={isRowExpanded ? Icons.ChevronUp : Icons.ChevronDown} />
            </IconButton>
          </DataGridCell>
        )}
        {visibleCells}
      </tr>
      {enableExpandableRow && (
        <tr className={`${classes["row"]} ${isRowExpanded ? classes["border-drawer"] : ""}`}>
          <td colSpan={visibleCells.length + 1}>
            <DataGridDrawer
              id={drawerId}
              role="region"
              aria-labelledby={expandButtonId}
              isExpanded={isRowExpanded}
            >
              {expandableRowContent}
            </DataGridDrawer>
          </td>
        </tr>
      )}
      {enableNestedRow && renderRecurrentRow()}
    </Fragment>
  );
};

export const DataGridRow = React.forwardRef(DataGridRowComponent) as <T extends {}>(
  p: Props<T> & { ref?: Ref<HTMLTableRowElement> }
) => ReactElement;
