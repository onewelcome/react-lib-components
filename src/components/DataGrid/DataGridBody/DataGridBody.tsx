/*
 * Copyright 2022 OneWelcome B.V.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

import React, { ComponentPropsWithRef, ReactElement, Ref } from "react";
import { HeaderCell } from "../datagrid.interfaces";
import { DataGridCell } from "./DataGridCell/DataGridCell";
import { DataGridRow } from "./DataGridRow/DataGridRow";
import classes from "./DataGridBody.module.scss";
import { Typography } from "../../Typography/Typography";

export interface Props<T> extends Omit<ComponentPropsWithRef<"tbody">, "children"> {
  children: ({ item, index }: { item: T; index: number }) => ReactElement;
  data?: T[];
  nestedRowConfig?: {
    nestedItemsKey?: keyof T;
  };
  headers: HeaderCell[];
  isLoading?: boolean;
  disableContextMenuColumn?: boolean;
  searchValue?: string;
  emptyLabel?: string;
  spacing?: React.CSSProperties;
}

const skeletonLoadingRows = 9;

const DataGridBodyInner = <T extends {}>(
  {
    children,
    data,
    headers,
    nestedRowConfig,
    isLoading,
    disableContextMenuColumn,
    emptyLabel,
    spacing,
    searchValue,
    ...rest
  }: Props<T>,
  ref: Ref<HTMLTableSectionElement>
) => {
  const renderContent = () => {
    const visibleColumns = headers.filter(header => !header.hidden).length;
    if (isLoading) {
      return Array.from(Array(skeletonLoadingRows)).map((_, rowIdx) => (
        <DataGridRow key={rowIdx} isLoading>
          {Array.from(Array(visibleColumns)).map((__, colIdx) => (
            <DataGridCell key={colIdx} isLoading></DataGridCell>
          ))}
          {!disableContextMenuColumn && <DataGridCell></DataGridCell>}
        </DataGridRow>
      ));
    }

    const emptyData = !data || data.length === 0;
    if (emptyData) {
      return (
        <tr>
          <td
            className={classes["empty"]}
            colSpan={visibleColumns + (disableContextMenuColumn ? 0 : 1)}
          >
            <Typography variant="body" spacing={{ margin: 0 }}>
              {emptyLabel}
            </Typography>
          </td>
        </tr>
      );
    }

    return data?.map((item, index) => {
      return React.cloneElement(children({ item, index }), {
        searchValue: searchValue,
        headers,
        spacing,
        item,
        disableContextMenuColumn,
        enableNestedRow: !!nestedRowConfig,
        nestedItemsKey: nestedRowConfig?.nestedItemsKey,
        rowTemplate: children,
        isLastChild: index + 1 === data.length
      });
    });
  };

  return (
    <tbody {...rest} ref={ref}>
      {renderContent()}
    </tbody>
  );
};

export const DataGridBody = React.forwardRef(DataGridBodyInner) as <T extends {}>(
  p: Props<T> & { ref?: Ref<HTMLTableSectionElement> }
) => ReactElement;
