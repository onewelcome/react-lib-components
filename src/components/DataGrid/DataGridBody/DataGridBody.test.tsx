/*
 * Copyright 2022 OneWelcome B.V.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

import React, { useEffect, useRef } from "react";
import { DataGridBody, Props } from "./DataGridBody";
import { render, getAllByRole } from "@testing-library/react";
import { DataGridRow } from "./DataGridRow/DataGridRow";
import { DataGridCell } from "./DataGridCell/DataGridCell";
import { ContextMenu } from "../../ContextMenu/ContextMenu";
import { IconButton } from "../../Button/IconButton";
import { Icon, Icons } from "../../Icon/Icon";
import { ContextMenuItem } from "../../ContextMenu/ContextMenuItem";

type DataType = { firstName: string; lastName: string; date: string };

const defaultParams: Props<DataType> = {
  children: ({ item }) => (
    <DataGridRow key={item.firstName}>
      <DataGridCell>{item.firstName}</DataGridCell>
      <DataGridCell>{item.lastName}</DataGridCell>
      <DataGridCell>{item.date}</DataGridCell>
      <DataGridCell>
        <ContextMenu
          id={`consent_menu_${item.firstName}`}
          trigger={
            <IconButton title={`Actions for ${item.firstName}`} color="default">
              <Icon icon={Icons.EllipsisAlt} />
            </IconButton>
          }
        >
          <ContextMenuItem>Item 1</ContextMenuItem>
        </ContextMenu>
      </DataGridCell>
    </DataGridRow>
  ),
  headers: [
    { name: "firstName", headline: "First name" },
    { name: "lastName", headline: "Last name" },
    { name: "date", headline: "Date" }
  ],
  data: [
    { firstName: "Paweł", lastName: "Napieracz", date: "12.12.1990" },
    { firstName: "Michał", lastName: "Górski", date: "12.12.1994" },
    { firstName: "Daniel", lastName: "Velden", date: "12.12.199x" },
    { firstName: "Jasha", lastName: "Joachimsthal", date: "12.12.198x" }
  ]
};

const createDataGridBody = (params?: (defaultParams: Props<DataType>) => Props<DataType>) => {
  let parameters = defaultParams;
  if (params) {
    parameters = params(defaultParams);
  }
  const container = document.createElement("table");
  const queries = render(<DataGridBody {...parameters} data-testid="dataGridBody" />, {
    container
  });
  const dataGridBody = queries.getByTestId("dataGridBody");

  return {
    ...queries,
    dataGridBody
  };
};

describe("DataGridBody should render", () => {
  it("renders without crashing", () => {
    const { dataGridBody } = createDataGridBody();

    expect(dataGridBody).toBeDefined();
    const rows = getAllByRole(dataGridBody, "row");

    expect(rows).toMatchSnapshot();
  });

  it("renders loading state with only visible columns", () => {
    const headers = [...defaultParams.headers];
    headers[0] = { ...headers[0], hidden: true };
    const { dataGridBody } = createDataGridBody(params => ({ ...params, headers }));

    expect(dataGridBody).toBeDefined();
    const rows = getAllByRole(dataGridBody, "row");
    const firstRowCells = getAllByRole(rows[0], "cell");
    expect(firstRowCells).toHaveLength(3);
    expect(firstRowCells[0]).toHaveTextContent(defaultParams.data?.[0].lastName!);
  });

  it("renders empty state", () => {
    const emptyLabel = "emptyLabel";
    const { dataGridBody } = createDataGridBody(params => ({
      ...params,
      emptyLabel,
      data: undefined
    }));

    const rows = getAllByRole(dataGridBody, "row");
    const firstRowCells = getAllByRole(rows[0], "cell");
    expect(firstRowCells).toHaveLength(1);
    expect(firstRowCells[0]).toHaveAttribute("colspan", `${defaultParams.headers.length + 1}`);
    expect(firstRowCells[0]).toHaveTextContent(emptyLabel);
  });
});

describe("ref should work", () => {
  it("should give back the proper data prop, this also checks if the component propagates ...rest properly", () => {
    const ExampleComponent = ({
      propagateRef
    }: {
      propagateRef: (ref: React.RefObject<HTMLElement>) => void;
    }) => {
      const ref = useRef(null);

      useEffect(() => {
        propagateRef(ref);
      }, [ref]);

      return <DataGridBody {...defaultParams} data-ref="testing" ref={ref} />;
    };

    const refCheck = (ref: React.RefObject<HTMLElement>) => {
      expect(ref.current).toHaveAttribute("data-ref", "testing");
    };

    const container = document.createElement("table");
    render(<ExampleComponent propagateRef={refCheck} />, { container });
  });
});
