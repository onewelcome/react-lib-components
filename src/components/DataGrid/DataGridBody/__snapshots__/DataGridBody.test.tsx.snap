// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataGridBody should render renders without crashing 1`] = `
[
  <tr
    class="row"
  >
    <td
      class="cell "
    >
      <span
        class="typography_style_body "
      >
        Paweł
      </span>
    </td>
    <td
      class="cell "
    >
      <span
        class="typography_style_body "
      >
        Napieracz
      </span>
    </td>
    <td
      class="cell "
    >
      <span
        class="typography_style_body "
      >
        12.12.1990
      </span>
    </td>
    <td
      class="cell "
    >
      <span
        class="typography_style_body "
      >
        <div
          class="context-menu"
        >
          <button
            aria-controls="consent_menu_Paweł-menu"
            aria-expanded="false"
            aria-haspopup="true"
            class="button  icon-button text primary button-m"
            id="consent_menu_Paweł"
            tabindex="0"
            title="Actions for Paweł"
            type="button"
          >
            <span
              aria-hidden="true"
              class="icon icon-ellipsis-alt "
              data-icon="true"
            />
            <span
              class="sr-only"
            >
              Actions for Paweł
            </span>
          </button>
        </div>
      </span>
    </td>
  </tr>,
  <tr
    class="row"
  >
    <td
      class="cell "
    >
      <span
        class="typography_style_body "
      >
        Michał
      </span>
    </td>
    <td
      class="cell "
    >
      <span
        class="typography_style_body "
      >
        Górski
      </span>
    </td>
    <td
      class="cell "
    >
      <span
        class="typography_style_body "
      >
        12.12.1994
      </span>
    </td>
    <td
      class="cell "
    >
      <span
        class="typography_style_body "
      >
        <div
          class="context-menu"
        >
          <button
            aria-controls="consent_menu_Michał-menu"
            aria-expanded="false"
            aria-haspopup="true"
            class="button  icon-button text primary button-m"
            id="consent_menu_Michał"
            tabindex="0"
            title="Actions for Michał"
            type="button"
          >
            <span
              aria-hidden="true"
              class="icon icon-ellipsis-alt "
              data-icon="true"
            />
            <span
              class="sr-only"
            >
              Actions for Michał
            </span>
          </button>
        </div>
      </span>
    </td>
  </tr>,
  <tr
    class="row"
  >
    <td
      class="cell "
    >
      <span
        class="typography_style_body "
      >
        Daniel
      </span>
    </td>
    <td
      class="cell "
    >
      <span
        class="typography_style_body "
      >
        Velden
      </span>
    </td>
    <td
      class="cell "
    >
      <span
        class="typography_style_body "
      >
        12.12.199x
      </span>
    </td>
    <td
      class="cell "
    >
      <span
        class="typography_style_body "
      >
        <div
          class="context-menu"
        >
          <button
            aria-controls="consent_menu_Daniel-menu"
            aria-expanded="false"
            aria-haspopup="true"
            class="button  icon-button text primary button-m"
            id="consent_menu_Daniel"
            tabindex="0"
            title="Actions for Daniel"
            type="button"
          >
            <span
              aria-hidden="true"
              class="icon icon-ellipsis-alt "
              data-icon="true"
            />
            <span
              class="sr-only"
            >
              Actions for Daniel
            </span>
          </button>
        </div>
      </span>
    </td>
  </tr>,
  <tr
    class="row"
  >
    <td
      class="cell "
    >
      <span
        class="typography_style_body "
      >
        Jasha
      </span>
    </td>
    <td
      class="cell "
    >
      <span
        class="typography_style_body "
      >
        Joachimsthal
      </span>
    </td>
    <td
      class="cell "
    >
      <span
        class="typography_style_body "
      >
        12.12.198x
      </span>
    </td>
    <td
      class="cell "
    >
      <span
        class="typography_style_body "
      >
        <div
          class="context-menu"
        >
          <button
            aria-controls="consent_menu_Jasha-menu"
            aria-expanded="false"
            aria-haspopup="true"
            class="button  icon-button text primary button-m"
            id="consent_menu_Jasha"
            tabindex="0"
            title="Actions for Jasha"
            type="button"
          >
            <span
              aria-hidden="true"
              class="icon icon-ellipsis-alt "
              data-icon="true"
            />
            <span
              class="sr-only"
            >
              Actions for Jasha
            </span>
          </button>
        </div>
      </span>
    </td>
  </tr>,
]
`;
